File: app/src/main/java/de/uni_marburg/mathematik/ds/serval/controller/view_holders/MeasurementsViewHolder.java
package de.uni_marburg.mathematik.ds.serval.controller.view_holders;

import android.content.Intent;
import android.os.Build.VERSION;
import android.os.Build.VERSION_CODES;
import android.support.annotation.LayoutRes;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.Locale;
import java.util.StringJoiner;

import butterknife.BindView;
import de.uni_marburg.mathematik.ds.serval.R;
import de.uni_marburg.mathematik.ds.serval.model.event.Measurement;


/**
 * {@link BaseViewHolder ViewHolder} for {@link Measurement measurements}.
 */
public class MeasurementsViewHolder extends BaseViewHolder<Measurement> {
    
    @BindView(R.id.measurement_type)
    TextView measurementType;
    @BindView(R.id.measurement_value)
    TextView measurementValue;
    @BindView(R.id.measurement_icon)
    ImageView measurementIcon;
    @BindView(R.id.explore)
    Button explore;
    @BindView(R.id.share)
    Button share;
    
    /**
     * Creates a new ViewHolder.
     *
     * @param parent       Parent ViewGroup
     * @param itemLayoutId Layout resource id
     */
    public MeasurementsViewHolder(ViewGroup parent, @LayoutRes int itemLayoutId) {
        super(parent, itemLayoutId);
        explore.setOnClickListener(this);
        share.setOnClickListener(this);
    }
    
    @Override
    protected void onBind(Measurement measurement, int position) {
        data = measurement;
        setupViews();
    }
    
    @Override
    protected void onClick(View view, Measurement measurement) {
        switch (view.getId()) {
            case R.id.explore:
                // TODO Explain measurement type
                break;
            case R.id.share:
                Intent shareIntent = new Intent();
                shareIntent.setAction(Intent.ACTION_SEND);
                String text = data.getType().toString() + " " +
                              context.getString(R.string.measurement).toLowerCase() +
                              context.getString(R.string.with_value) + data.getValue();
                shareIntent.putExtra(Intent.EXTRA_TEXT, text);
                shareIntent.setType(context.getString(R.string.intent_type_text_plain));
                context.startActivity(Intent.createChooser(
                        shareIntent,
                        context.getResources().getText(R.string.chooser_title_share_measurement)
                ));
                break;
        }
    }
    
    
    @Override
    protected boolean onLongClick(View view, Measurement data) {
        return false;
    }
    
    /**
     * Sets up all views.
     */
    private void setupViews() {
        measurementType.setText(data.getType().toString());
        String value;
        int resId;
        switch (data.getType()) {
            case PRECIPITATION:
                value = context.getString(R.string.measurement_value_precipitation);
                resId = R.drawable.precipitation;
                break;
            case RADIATION:
                value = context.getString(R.string.measurement_value_radiation);
                resId = R.drawable.radiation;
                break;
            case TEMPERATURE:
                value = context.getString(R.string.measurement_value_temperature);
                resId = R.drawable.temperature;
                break;
            case WIND:
                value = context.getString(R.string.measurement_value_wind);
                resId = R.drawable.wind;
                break;
            default:
                value = context.getString(R.string.measurement_value_placeholder);
                resId = R.drawable.dissatisfied;
                break;
        }
        measurementValue.setText(String.format(
                Locale.getDefault(),
                value,
                data.getValue()
        ));
        measurementIcon.setImageResource(resId);
    }
}

File: app/src/main/java/de/uni_marburg/mathematik/ds/serval/controller/view_holders/MeasurementsViewHolder.kt
package de.uni_marburg.mathematik.ds.serval.controller.view_holders

import android.content.Intent
import android.support.v7.widget.RecyclerView
import android.view.View
import de.uni_marburg.mathematik.ds.serval.R
import de.uni_marburg.mathematik.ds.serval.model.Measurement
import de.uni_marburg.mathematik.ds.serval.model.MeasurementType
import kotlinx.android.synthetic.main.measurement_row.view.*
import java.util.*


/** [View holder][RecyclerView.ViewHolder] for [measurements][Measurement] */
class MeasurementsViewHolder(itemview: View) :
        RecyclerView.ViewHolder(itemview),
        View.OnClickListener,
        View.OnLongClickListener {

    lateinit var measurement: Measurement

    fun performBind(measurement: Measurement) {
        this.measurement = measurement
        setupViews()
    }

    override fun onClick(view: View) {
        when (view.id) {
            R.id.explore -> TODO()
            R.id.share -> {
                val shareIntent = Intent()
                shareIntent.action = Intent.ACTION_SEND
                val text = measurement.type.toString() + " " +
                        itemView.context.getString(R.string.measurement).toLowerCase() +
                        itemView.context.getString(R.string.with_value) + measurement.value
                shareIntent.putExtra(Intent.EXTRA_TEXT, text)
                shareIntent.type = itemView.context.getString(R.string.intent_type_text_plain)
                itemView.context.startActivity(Intent.createChooser(
                        shareIntent,
                        itemView.context.resources.getText(R.string.chooser_title_share_measurement)
                ))
            }
        }
    }

    override fun onLongClick(view: View): Boolean = false

    /** Sets up all views. */
    private fun setupViews() {
        val value: String
        val resId: Int
        when (measurement.type) {
            MeasurementType.PRECIPITATION -> {
                value = itemView.context.getString(R.string.measurement_value_precipitation)
                resId = R.drawable.precipitation
            }
            MeasurementType.RADIATION -> {
                value = itemView.context.getString(R.string.measurement_value_radiation)
                resId = R.drawable.radiation
            }
            MeasurementType.TEMPERATURE -> {
                value = itemView.context.getString(R.string.measurement_value_temperature)
                resId = R.drawable.temperature
            }
            MeasurementType.WIND -> {
                value = itemView.context.getString(R.string.measurement_value_wind)
                resId = R.drawable.wind
            }
        }

        itemView.measurement_type.text = measurement.type.toString()
        itemView.measurement_value.text = String.format(
                Locale.getDefault(),
                value,
                measurement.value
        )
        itemView.measurement_icon.setImageResource(resId)
    }
}
